astropy/table/table.py

```diff
@@ -595,6 +595,13 @@ def __init__(self, data=None, masked=False, names=None, dtype=None,
                 except Exception:
                     raise ValueError('dtype was specified but could not be '
                                      'parsed for column names')
+            elif dtype is not None:
```

Comment:
My idea was actually to move this outside of the `data is None` block to
```
        if names is None:
            names = default_names or [None] * n_cols
        if dtype is None:
            dtype = [None] * n_cols
        elif isinstance(dtype, np.dtype):  
            # Names supplied along with structured-array dtype. Convert to
            # a list of dtypes for later use. If dtype is already a list of
            # dtype-specifiers e.g. ['f4', 'i4'] then skip this.
            dtype = [dtype[name] for name in dtype.names]
```
so this would also allow changing the original `data.dtype` from an `np.dtype` (rather than just a list of dtypes).